Feature/aztec 2d

mneuber opened the pull request:
This is a continuation ( and finalization? ) of the aztec codec implementation. 
Tests for Detector/Decoder/BlackBox are all passing.
3 of 18 test for the encoder failing, due to some encoding limitations in js with Textencoder class ( can only encode in utf-8 ). But there could be some more issues that cause the failing.
Some bugs were caused by integer divisions which are truncated in JAVA but not in js. Math.trunc() is the best choice to match this behaviour in my eyes.
![Screenshot 2020-04-24 at 18 09 23](https://user-images.githubusercontent.com/14347067/80234219-0f4a6800-8658-11ea-875e-22f1b704c8fa.png)
Unfortunately prettier messed up some git diffs, hope that's gonna be fine for you.

odahcam commented:
In case of trancation, I recommend adding a comment about the action and adding the `Math.trunc` to match Java's behavior.

The UTF-8 issue, we can easily workaround with text-encoder, which is installed with the NPM packages.

I'd like we fix the prettier mess, at least to match the library's style.

odahcam commented:
You can add these lines before the test to add support for decoding in other encodings not natively supported by Node:
https://github.com/zxing-js/library/blob/70e9c3a1b5d9c2012bf059a5a3fa466b9ac376c6/src/test/core/pdf417/PDF417BlackBox.2.spec.ts#L22-L25

mneuber commented:
yes let me clean up the prettier mess in the next commits and I can add comments where truncation was necessary. We could have add a descriptive wrapper function like `intDivision`, which is doing the job, I'm fine with either way. 

I might miss something but decoding in different encodings works fine. But encoding in different encodings seems not supported yet, as the comment here states: https://github.com/mneuber/library/blob/2a623a92ffd566d4cb9950841fe924484bf1eb60/src/core/util/StringEncoding.ts#L66-L67

odahcam commented:
Okay, I like the `intDivision` idea, we could create a function like that or even `truncDivision` to be explicit about what happens is not rounding. Also, for non UTF-8 encodings you may need to use a custom text **encoder** as well: `StringEncoding.customEncoder = (s, e) => new TextEncoder(s, { NONSTANDARD_allowLegacyEncoding: true }).encode(s);`.

mneuber commented:
prettier changes are gone, _Files changed_ is pretty clean now. 
I added a static function to Integer class called truncDivision and replaced Math.trunc by that. 
Added custom Textencoder in encoder tests, but the same 3 tests still failing. At least we have now an error thrown, which would be starting point to solve them.

odahcam commented on src/test/core/aztec/encoder/EncoderTest.spec.ts:
```diff
@@ -776,8 +779,6 @@ describe('EncoderTest', () => {
   }
 
   function testHighLevelEncodeString(s: string, expectedBits: string | number) {
-    const jhu = StringUtils.getBytes(s, StandardCharsets.ISO_8859_1);
```
Could you please comment the line instead of removing it? It may serve as reference as future debugging, like: 
```typescript
// TYPESCRIPTPORT: removed follwoing line because of...
// const jhu = StringUtils.getBytes(s, StandardCharsets.ISO_8859_1);
```

odahcam commented on src/core/util/Integer.ts:
```diff
@@ -101,6 +101,10 @@ export default class Integer {
   }
 
+  public static truncDivision(i: number): number {
+    return Math.trunc(i);
```
Oh I thought this would be a method for performing the division, like:

```typescript
  public static truncDivision(a: number, b: number): number {
    const result = a / b;
    return Math.trunc(result);
  }
```

I see no reason for just wrapping `Math.trunc` inside another function that will aggregate no functionality.

mneuber replied:
yep, I thought about that while doing. Ok, I will change that one.

odahcam commented on src/test/core/aztec/decoder/Decoder.spec.ts:
```diff
@@ -167,14 +163,18 @@ describe('DecoderTest', () => {
+            true];
+        let bool16: boolean[] = [
+            false, true, true, false, false, false, true, true,
+            true, true, false, false, false, false, false, true];
```
I couldn't understand the logic behind these new linebreaks around the code. Did you do these line breaks or was it a result from tslint or prettier-like tools? Just curious.

odahcam commented on src/core/aztec/detector/Detector.ts:
```diff
@@ -143,7 +143,7 @@ export default class Detector {
             this.sampleLine(bullsEyeCorners[1], bullsEyeCorners[2], length), // Bottom
             this.sampleLine(bullsEyeCorners[2], bullsEyeCorners[3], length), // Left side
-            this.sampleLine(bullsEyeCorners[3], bullsEyeCorners[0], length), // Top
+            this.sampleLine(bullsEyeCorners[3], bullsEyeCorners[0], length)  // Top
```
Trailing comma should be used, please don't bother removing it and please forgive me if and linter rules says the opposite, in that case they would be wrong.

odahcam commented on .prettierrc:
```diff
@@ -1,6 +1,4 @@
 {
     "singleQuote": true,
-    "tabWidth": 4,
```
We do want to use tab width of 2, thanks for removing it. Although I have to ask you to not fix (set it to 2) this setting in this PR as it would generate too many out of context code for reviewing. Removing is great, thanks.

odahcam commented on src/core/aztec/decoder/Decoder.ts:
```diff
@@ -41,149 +43,33 @@ enum Table {
 
-    private static DIGIT_TABLE: string[] = ['CTRL_PS', ' ', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ',', '.', 'CTRL_UL', 'CTRL_US'];
+    private static DIGIT_TABLE: string[] = [
+        'CTRL_PS', ' ', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ',', '.', 'CTRL_UL', 'CTRL_US'
```
Oh @prettier, you do awful things sometimes.

odahcam commented on src/core/aztec/decoder/Decoder.ts:
```diff
@@ -346,7 +231,7 @@ export default class Decoder {
 
         let dataWords: Int32Array = new Int32Array(numCodewords);
-        for (let i = 0; i < numCodewords; i++, offset += codewordSize) {
+        for (let i = 0; i < numCodewords; i++ , offset += codewordSize) {
```
God why am I annoyed by this space. Damn space.

odahcam commented on src/core/aztec/decoder/Decoder.ts:
```diff
@@ -232,7 +117,7 @@ export default class Decoder {
                 for (let charCount = 0; charCount < length; charCount++) {
                     if (endIndex - index < 8) {
-                        index = endIndex; // Force outer loop to exit
+                        index = endIndex;  // Force outer loop to exit
```
If there's some setting for double spaces for inserting comments at the EOL, we should change it to single space to match Java.

odahcam commented on .vscode/launch.json:
```diff
@@ -116,6 +116,24 @@
+    {
+      "type": "node",
+      "request": "launch",
+      "name": "Aztec 2D Tests - ts-node",
```
Is this faster than webpack-mocha?

mneuber replied:
I struggled to debug with mocha-webpack, either the source maps didn't match the typescript sources or it just didn't stop at the break points. mocha-webpack seems not maintained anymore since longer time. But it could be my unknowing in this topic.
This setup did just work fine for.

odahcam replied:
Nice to know that, it was just the opposite for me.

odahcam commented on src/core/aztec/encoder/CharMap.ts:
```diff
@@ -5,26 +5,30 @@ import StringUtils from '../../common/StringUtils';
-  CHAR_MAP[C.MODE_DIGIT][','] = 12;
-  CHAR_MAP[C.MODE_DIGIT]['.'] = 13;
+  CHAR_MAP[C.MODE_DIGIT][commaCharCode] = 12;
+  CHAR_MAP[C.MODE_DIGIT][pointCharCode] = 13;
```
As much as I loved this change, I would ask you to not improve much the Java code, I would like to keep it as close as possible at least until we release v1.0.0. But this is fine.

mneuber replied:
ok, it was really a bug fix as we compare char codes on some critical places.

mneuber replied:
In Java its an int array as well, it just auto converts char to int, we need to do it manually.

odahcam replied:
No problem, code looks nice.

odahcam commented on src/test/resources/blackbox/aztec-1/Historico.txt:
```diff
@@ -0,0 +1 @@
+Histórico
```
🇧🇷 🇵🇹

mneuber replied:
one of the strings which failed in the beginning ^^

odahcam replied:
I must admit these accents in our language doesn't make our lives easier hehe.

odahcam commented:
I've fixed one of the failing tests but I'm struggling with the `testHighLevelEncodeBinary`, where 
`StringUtils` can't encode the sequence `'N\0\u0080 A'` of characters, giving the error: 

```
TypeError: The code point 128 could not be encoded.
```

The character in question is this one: [`\u0080`](https://www.fileformat.info/info/unicode/char/0080/index.htm), which has the decimal value of `128` or hexadecimal `80` (the same in Java).

I tried switching from ISO 8859-1 encoding to UTF-8 and ISO 8859-15 (and it encodes properly) for the test method in here:

https://github.com/mneuber/library/blob/bf7bdd6e62c1666ff93f6df3544e08d4cc48fc51/src/test/core/aztec/encoder/EncoderTest.spec.ts#L789-L792

But the results still fails. From this point I started to think there's another error with the encoding inside the Encoder itself, I just couldn't find it yet, but I wanted to make an update on this.

mneuber commented:
How is the review progressing? Do you miss some requested changes? The extra launcher is not really necessary, just wanted to offer it for debugging purposes.

odahcam commented:
Oh sorry, I thought it was clear we had to fix the still failing tests. Unfortunately I can't merge the PR into the master with decoding issues. Still, I can merge it in the feature branch, but I would not publish it until all tests pass.

mneuber commented:
No problem, at the end the title and description of this PR indicates that this is feature complete.
> Unfortunately I can't merge the PR into the master with decoding issues.
Ok, but I guess you mean only encoding issues.

I would be glad if detector/decoder would get merged. For the encoder we could create a new PR.

Beside of this, some observations about the encoding issues:
In our  custom encoder (@sinonjs) we find that piece of code:

      // 2. If code point is an ASCII code point, return a byte whose
      // value is code point.
      if (isASCIICodePoint(code_point))
        return code_point;

     // 3. Let pointer be the index pointer for code point in index
      // single-byte.
      var pointer = indexPointerFor(code_point, index);

      // 4. If pointer is null, return error with code point.
      if (pointer === null)
        encoderError(code_point);

I think here is our error thrown and the point is: 
isASCIICodePoint is only checking for the range 0x00 to 0x7F, while our character in question has a byte code of 0X80, hence it is trying to get the index for the charcode. But in ISO8859-1 there is no char defined for index 128. And the error is thrown.

In Javas StringEncoding class, I found that price of code:

`private static byte[] encode8859_1(byte coder, byte[] val, boolean doReplace) {
        if (coder == LATIN1) {
            return Arrays.copyOf(val, val.length);
        }`

Here it is checking for Latin1 and if that refers to https://en.wikipedia.org/wiki/Latin-1_Supplement_(Unicode_block) it includes our charCode and just giving back the byte code.
Our implementation is refusing that as it just checking for the ASCII range. 

This is really just an observation, but would explain why in our implementation it fails, while in Java it works. 
But this all I can provide for now.

odahcam commented:
> Ok, but I guess you mean only encoding issues.

Yes, sorry.

Good observations tho, it's a strange behavior indeed. Here [in this table](https://www.ic.unicamp.br/~stolfi/EXPORT/www/ISO-8859-1-Encoding.html) it says the 128 (x080) is not used at all, maybe it changed on some point in time.

I will then merge this PR into the feature branch and publish it without the encoder when possible.

mneuber commented:
Be aware that the encoder is used in the test suit of the detector. 
Maybe we can label the encoder as "beta" or whatever would be common here.
Thing is that the encoder should work for the most of the daily cases.

username1565 commented:
@mneuber, I don't understand typescript,
but as you can see here: https://unicode-table.com/en/
ASCII-characters, and Latin-1-characters (iso-8859-1) -
this is the part of Unicode characters table, and Unicode have the backward compatibility with this both encoding.

Latin-1 can encode all 256 bytes to string, and back.
This is a reversive encoding, and this can encode all bytes to string, and decode it back.
But ISO-8859-1 encoding (LATIN-1) contains the special characters,
and string can not be copied, sometimes.
This is the special-symbol with code 0x80 (128 decimal): https://unicode-table.com/en/0080/
and in JavaScript this is working:
```
console.log('\x80'.charCodeAt(0)); //128
```
also, this symbol can be copied:
```
console.log('\x80') // "", and this is copyable.
```

mneuber commented:
@username1565 @odahcam 
seems like the bug occurs only in nodejs environment.
we can break it down to this piece of code ( es6 ):
```
import { TextEncoder } from "@sinonjs/text-encoding";
const byte = new TextEncoder("ISO-8859-1", { NONSTANDARD_allowLegacyEncoding: true}).encode("\u0080");
console.log(byte); //{0: 194, 1: 128}
```

Just tested in a sandbox and it works:
https://codesandbox.io/s/testencoder-test-i7v04

When we execute the same piece of code in nodejs in our test suite, we receive this error:

```
TypeError: The code point 128 could not be encoded.
    at encoderError (node_modules/@sinonjs/text-encoding/lib/encoding.js:285:11)
    at SingleByteEncoder.handler (node_modules/@sinonjs/text-encoding/lib/encoding.js:1622:9)
    at TextEncoder.encode (node_modules/@sinonjs/text-encoding/lib/encoding.js:1299:30)
```

username1565 commented:
@mneuber, this is the problem of TextEncoder.
By default, as you can see [here](https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder), this can return only "utf-8"-bytes, after encoding:
```
Always returns "utf-8".
```

See how I did resolve this here: https://github.com/sinonjs/text-encoding/pull/2/commits/c1193bd1cc3fe52f826256a6b4985869526c50b9

1. Add "iso-8859-1", (and keywods, like `latin1`, `latin-1`, etc.) into `encoding.js`.
2. Add `iso-8859-1`, into `encoding-indexes.js`.
3. Override `TextEncoder` and `TextDecoder` which was been already defined in browser.
4. Test this, and use that two functions, if `TextEncoder` and `TextDecoder` - not working correctly.

mneuber commented:
@username1565 
ah ok. So this bug should be fixed with one of the next versions, if I am right.

username1565 commented:
@mneuber, anyway, you can fix this yourself, because all is open-source.
The following string in your code:
```
import { TextEncoder } from "@sinonjs/text-encoding";
```
can be used to loading already fixed `TextEncoder`/`TextDecoder`, and override the default, with this loaded code.

odahcam commented:
> He not answered me from October 2019. Maybe he lost the password.

God this package is disturbing for so long that I will have to fork it myself and force everyone here to 2FA.

Just kidding.

But I would gladdly fork this piece of nightmare to @zxing-js and proper maintain it if you'd like. I think maybe that's a good idea since everyone that touches it alone seems to loose access to it's account. Maybe it's force is so strong it encodes users passwords for good.

Let me know if forking in here is an option (I would you maintainer if you want, would be good to have some company around here) and I will fork it so we can proper maintain and publish the package.

username1565 commented:
@odahcam, if you don't want to override `TextEncoder` and `TextDecoder` just to privide Latin1-compatibility, you can just use [that two functions](https://github.com/username1565/text-encoding/blob/c1193bd1cc3fe52f826256a6b4985869526c50b9/lib/index.html#L93-L107) to encode-decode `iso-8859-1 (Latin1)`, instead `TextEncoder/TextDecoder`
This will be better and easy to re-write this on `TypeScript` from `JavaScript`.

_Have a nice day_

odahcam commented:
Thanks for all your help @username1565 ! I think I'll fork this package and update it's code with yours and @sinonjs's code so we can improve our Node.js compatibility. You're welcome to join if you like.
